util/fpga/cw_spiflash.py

```diff
@@ -0,0 +1,273 @@
+# Copyright lowRISC contributors.
```

Comment:
@cfrantz This looks prime for ottoolization!

Reply:
This appears to be a python implementation of the current bootstrap protocol.

I have no specific objection to this except to mention that there is already a C++ based version of the bootstrap protocol in sw/host/spiflash with some abstraction over the SPI interface such that support for different hardware interfaces (ie: FTDI chips/whatever) can be added.

Could that pre-existing code work for you?

There is a longer-term plan to unify the hardware debug interfaces to FPGA boards under a single tool, however that effort is not underway yet (and shouldn't block progress on this PR).

Reply:
I can't judge on the feasibility, required effort of adding support for the ChipWhisperer boards to `sw/host`spiflash`. The main difference is that the CW boards don't have an FTDI but a SAM3X/U that handles SPI communication with OpenTitan.

In the early days of our SCA experiments we have been using a separate FTDI cable together with `sw/host/spiflash` for loading binaries into OpenTitan on the CW305 to avoid going through the SAM3U. It worked but the FTDI cable injected heavy noise into the power traces (even when idling) which is very bad for the experiments. Thankfully @alphan went ahead and routed SPI through the SAM3X/U to get rid of these issues. ChipWhisperer provides the Python library to handle SPI over the SAM3X/U, don't know if something similar is available in C++. Maybe @alphan can quickly comment.

Anyway, I will add a point to the tracking issue to investigate that.

Reply:
As @vogelpi mentioned, this implements the current bootstrap protocol using SAM3U's [gpio mode](https://chipwhisperer.readthedocs.io/en/latest/api.html#cw305-sam3u-io-control), i.e. `host -> SAM3U -> CW305` instead of `host -> ftdi -> CW305`. 

That being said, I'm not sure if it's a good idea to keep this script in the opentitan repo since it only exists to reduce noise. It would be great if we could use CW310 in opentitan using both ROMs, i.e. `boot_rom` and `mask_rom`, and existing tools and keep this script in ot-sca. I'm not sure if we will have issues with pins though. @vogelpi WDYT?

Reply:
> I can't judge on the feasibility, required effort of adding support for the ChipWhisperer boards to `spiflash`. The main difference is that the CW boards don't have an FTDI but a SAM3X/U that handles SPI communication with OpenTitan.

`spiflash` already supports programming using an FTDI cable (of course, CW310's pins are routed differently now to be able to use SAM3U) so that can be an option using existing code and hardware. Removing the FTDI cable requires talking to CW310 endpoints over USB as this script does. This script uses `chipwhisperer` (which we need anyway for loading bitstreams, configuring CW310, and occasionally updating SAM3U's firmware) to do that. 

> ChipWhisperer provides the Python library to handle SPI over the SAM3X/U, don't know if something similar is available in C++.

I don't think so.

On a second thought, it may not be a bad idea to use this script in opentitan too since it doesn't require an FTDI cable and simplifies our setup. It would be great if we could enhance this to support the new bootstrap protocol that's in the works.